### ISSUE TITLE
Should borg assume that it's okay to store unencrypted metadata information of the repositories in the cache without asking?

### ISSUE THREAD

**Issue by finefoot** [borgbackup/borg#5041]
This question came up in another issue and I'm just posting it separately to improve discussing the topic.

When I first started using borg for a simple backup task, I didn't know about the cache or what it does. I mean, *now* I know about it and the fact that it stores unencrypted metadata. I could have read the internals section in the docs before using borg, but I just didn't.

Does borg assume that its users know that the cache stores unencrypted metadata of the repositories? Does borg assume that its users read the internals section?

Currently, borg assumes

> that the cache is being stored on the very same system which also contains the original files which are being backed up. So someone with access to the cache files would also have access the the original files anyway.

(Quote from the docs FAQ.)

However, I'm still not sure if this should happen silently without informing the user. Additionally, one might use borg to back up an external hard drive: without knowingly changing the cache directory, the cache won't be on the same system as the original files. (Cache in possibly unencrypted home folder, containing information about the files on the backed up encrypted external hard drive.)

**Comment by finefoot** [comment]
How about showing a "warning" asking the user when no cache is found (which would generally be only once per repository) if they want to continue with `--no-files-cache` or if they are okay with creating a cache? If there already is a cache found for the affected repository, it would be safe to assume that it's okay to keep using it.

**Comment by finefoot** [comment]
To summarize the discussion off of IRC, a compromise would be to show a non-blocking notice when a new cache is being created: `A cache for this repository has been created at {}. Please make sure that you have read the cache security notes in the docs.`

**Comment by fantasya-pbem** [comment]
I just checked when a cache is created (at create or already at init) - it's at borg init. In init there is already some output and a password request, so it would make some sense to show a cache "warning" there (unless the password isn't provided by environment variable).

If between two create the cache disappears, there is currently some output about recreating the cache when in info mode (borg create -v). One could then also output the cache "warning". But a simple create without -v doesn't output anything (stays silent), so I would not change that if a cache has disappeared.

So these are possible places for a cache security notice:
- borg init -e <everything-but-none> /path/to/repo (no passphrase via envvar)
- borg init -v ...
- borg create -v ...

There may be more, but we see a pattern: Only output the notice when "in verbose mode".

**Comment by finefoot** [comment]
> But a simple create without -v doesn't output anything (stays silent), so I would not change that if a cache has disappeared.

I agree. For cache creation via `borg create` and not `borg init`, there is no need for the notice without `-v`. After all, the user must have used `--no-files-cache` before (which means they were aware of the cache existence) or has seen the notice at least once during `borg init`. That's what you meant for that case, too, yes?

> unless the password isn't provided by environment variable

Can you elaborate on that?

**Comment by fantasya-pbem** [comment]
> Can you elaborate on that?

Of course. AFAIK the env variable BORG_NEW_PASSPHRASE is respected by ``borg init`` and does not request the pw from the user. I don't remeber if init stays completely silent in that case, but we should not just output a cache notice then, too.

**Comment by finefoot** [comment]
> AFAIK the env variable BORG_NEW_PASSPHRASE is respected by borg init and does not request the pw from the user. I don't remeber if init stays completely silent in that case, but we should not just output a cache notice then, too.

Ah, I understand. First part is true, there is no input query for a password. But there is still output:

```
$ export BORG_PASSPHRASE=123
$ borg init -e repokey repo

By default repositories initialized with this version will produce security
errors if written to with an older version (up to and including Borg 1.0.8).

If you want to use these older versions, you can disable the check by running:
borg upgrade --disable-tam repo

See https://borgbackup.readthedocs.io/en/stable/changes.html#pre-1-0-9-manifest-spoofing-vulnerability for details about the security implications.

IMPORTANT: you will need both KEY AND PASSPHRASE to access this repo!
Use "borg key export" to export the key, optionally in printable format.
Write down the passphrase. Store both at safe place(s).

$ 
```

**Comment by finefoot** [comment]
That's why I thought, it would make sense to add a notice about the cache creation there, too.

**Comment by fantasya-pbem** [comment]
> That's why I thought, that it would make sense to add a notice about the cache creation there, too.

Absolutely!

**Comment by fantasya-pbem** [comment]
After some hours of sleep, I now think I changed my mind. Just did some ``borg init`` commands to compare the output of different encryption modes (no verbose option used, with BORG_NEW_PASSPHRASE set) to see what is there in the master branch.

- -e none
  - (no output)
- -e keyfile
  1. A note about security errors when writing with older versions.
  2. A note about ``borg upgrade --disable-tam``.
  3. A link to the docs regarding pre-1.0.9-manifest-spoofing-vulnerability.
  4. IMPORTANT: store key and passphrase, use ``borg key export``.
- -e repokey
  1. A note about security errors when writing with older versions.
  2. A note about ``borg upgrade --disable-tam``.
  3. A link to the docs regarding pre-1.0.9-manifest-spoofing-vulnerability.
  4. IMPORTANT: store key and passphrase, key should be backed up, use ``borg key export``.

That is a lot of information given every time I initialize a repo, event without giving a verbose option. For a new user who just inits the first repo ever it may be nice to know. But a brand new user should RTFM first.

A brand new user also won't do "weird" things like moving default config/cache directories to other places or switches back and forth between borg versions the first time, without looking into the manual how these things can be done.

So there is in general no need for all these notes in init, and for an additional cache security note.
I would rather support a task to get rid of all these information output, at least when no verbose option is given. Borg init should just init and stay silent, if no user interaction is required.

**Comment by finefoot** [comment]
> That is a lot of information given every time I initialize a repo

Does it happen that often that one initializes a new repository, though?

> But a brand new user should RTFM first.

I'm sorry, in my opinion that's just an utopian assumption. Even more, I actually did have a look at the docs before running borg but I must have skipped the essential paragraph about how borg stores unencrypted metadata in its cache.

Defensive design should be part of any sophisticated piece of software - especially regarding security-relevant aspects. For example, `borg init` *requires* the `--encryption` option. It could just take `none` as the default here, which means not having to dreadfully query a password. On the other hand, the requirement to specify a value safeguards a certain level of engagement with the available encryption modes.

> A brand new user also won't do "weird" things like moving default config/cache directories to other places

That doesn't take my example into consideration where the (default) cache location and backup source are not on the same filesystem. It might not be what *most* users do, but it's certainly not "weird".

> I would rather support a task to get rid of all these information output, at least when no verbose option is given. Borg init should just init and stay silent, if no user interaction is required.

At least that would be consistent behavior: Either show all relevant warnings/notices or don't show anything at all. In other words, don't show something about key backup while omitting the unencrypted cache.

**Comment by fantasya-pbem** [comment]
Another idea was to introduce a ``borg init --quiet`` option to suppress the info output that init normally prints. That would be a bit inconsistent to the other commands (suppress output vs. give some/more output). ``init`` is a somewhat special command in borg, but there is no real reason why this has to be. There may be use cases where init is a more often used command.

Or we could think of a change so that if Borg detects it has no ``~/.config/borg`` and ``~/.cache/borg`` yet (or whatever the user defined as config/cache), it could only then output some information for the would-be new user (save your password, cache is unencrypted, use the latest version and such).

**Comment by finefoot** [comment]
But do you really run `borg init` that often that it warrants a `--quiet` option or that it has to differentiate between completely new users and those who run `borg init` again?

**Comment by ThomasWaldmann** [comment]
we already have log levels, --info/warning/error.

On March 28, 2020 12:12:05 PM GMT+01:00, Thalian <notifications@github.com> wrote:
>Another idea was to introduce a ``borg init --quiet`` option to
>suppress the info output that init normally prints. That would be a bit
>inconsistent to the other commands (suppress output vs. give some/more
>output). ``init`` is a somewhat special command in borg, but there is
>no real reason why this has to be. There may be use cases where init is
>a more often used command.
>
>Or we could think of a change so that if Borg detects it has no
>``~/.config/borg`` and ``~/.cache/borg`` yet (or whatever the user
>defined as config/cache), it could only then output some information
>for the would-be new user (save your password, cache is unencrypted,
>use the latest version and such).
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/borgbackup/borg/issues/5041#issuecomment-605432308

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by fantasya-pbem** [comment]
> > But a brand new user should RTFM first.
> I'm sorry, in my opinion that's just an utopian assumption.
We cannot force them to do so, but to read the manual as a new user is nonetheless the most reasonable thing to do.

What about this compromise:

- Change ``borg init`` to stay silent if log level is above info (-v/--info not given).
- If it needs a password/runs interactively, it will ask for it (same as now).
- Add a --info note about unencrypted cache.
- Optional: If user forgot --encryption option, display an error (same as now) and add a note about unencrypted cache. This way a new user not familiar with the verbose options would get this info, too.
